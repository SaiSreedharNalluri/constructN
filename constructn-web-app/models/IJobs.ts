export interface IJobs {
  _id: string;
  project: Project;
  structure: Structure;
  createdBy: CreatedBy;
  assignedTo: string;
  status: string;
  captures?: CapturesEntityOrSelectedCapturesEntity[];
  selectedCaptures?: CapturesEntityOrSelectedCapturesEntity[];
  date: string;
  createdAt: string;
  updatedAt: string;
}
export interface Project {
  _id: string;
  type: string;
  name: string;
  company: string;
  email: string;
}
export interface Structure {
  _id: string;
  name: string;
  type: string;
  isExterior: boolean;
  project: string;
  parent: string;
  children?: null[] | null;
  location?: null[] | null;
}
export interface CreatedBy {
  _id: string;
  firstName: string;
  lastName: string;
  email: string;
  fullName: string;
}
export interface CapturesEntityOrSelectedCapturesEntity {
  _id: string;
  user: string;
  project: string;
  structure: string;
  captureDateTime: string;
  jobId: string;
  type: string;
  mode: string;
  isMediaUploaded: boolean;
  approvedStatus?: boolean;
  mediaPath: string;
}
