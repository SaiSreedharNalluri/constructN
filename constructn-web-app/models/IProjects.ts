export interface IProjects {
  _id: string;
  type: string;
  name: string;
  company?: null;
  email: string;
  contact: Contact;
  address: Address;
  location?: number[] | null;
  createdAt: string;
  updatedAt: string;
  jobsOpened?: number;
  LastUpdatedOn?: string;
  coverPhoto: string;
}
export interface Contact {
  code: string;
  number: number;
}
export interface Address {
  zipcode: string;
  city: string;
  state: string;
  country: string;
}
