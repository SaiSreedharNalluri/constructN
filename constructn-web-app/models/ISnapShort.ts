export interface ISnapShort {
  _id: string;
  date: string;
  progress: number;
  project: string;
  createdAt: string;
  updatedAt: string;
  __v: number;
  structure: string;
  status: string;
  visualization: [];
}
