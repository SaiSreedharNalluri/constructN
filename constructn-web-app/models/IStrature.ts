export interface IStrature {
  success: boolean;
  result?: IStratureEntity[] | null;
}
export interface IStratureEntity {
  _id: string;
  name: string;
  type: string;
  parent?: null;
  children?: ChildrenEntity[] | null;
}
export interface ChildrenEntity {
  _id: string;
  name: string;
  type: string;
  parent: string;
  children?: ChildrenEntity1[] | null;
}
export interface ChildrenEntity1 {
  _id: string;
  name: string;
  type: string;
  parent: string;
  children?: null[] | null;
}
