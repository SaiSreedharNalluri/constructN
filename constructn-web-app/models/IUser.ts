export interface IUser {
  _id: string;
  firstName: string;
  lastName: string;
  email: string;
  contact: Contact;
  dob: string;
  verified: boolean;
  createdAt: string;
  updatedAt: string;
  __v: number;
  fullName: string;
  age: number;
  token: string;
}
export interface Contact {
  code: string;
  number: number;
}
