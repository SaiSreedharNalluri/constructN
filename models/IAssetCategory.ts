export interface IAssetCategory {

    _id: string

    name: string

    project: string

    stages: IAssetStage[]

    shape: 'Polgon' | 'Line' | 'Circle' | 'Point'

    description?: string

    properties: any
    
}

export interface IAssetStage {

    _id: string

    name: string

    sequence: number

    color: string

    uom: string

    description?: string

}

export interface IAsset {

    _id: string

    name: string

    category: string | IAssetCategory

    progress: IAssetProgress

    progressSnapshot: IAssetProgress[]

    structure: string

    measurement: number

    measurementFactor: number

    shape: 'Polygon' | 'Polyline' | 'Circle'

    points: IAssetPoint[]

    description?: string

    createdBy: any

    properties: any
    
}

export interface IAssetProgress {

    stage: string | IAssetStage

    measurement: number

    uom: string

    date: Date,

    updatedBy: any

    updatedAt: Date

}

export interface IAssetPoint {

    x: number

    y: number

    r?: number

}

export const NOT_STARTED_STAGE: IAssetStage = { 
    
    name: 'NOT STARTED', 
    
    sequence: 0, 
    
    color: '#000080', 
    
    _id: 'NOT_STARTED', 
    
    uom: 'NA' 

}