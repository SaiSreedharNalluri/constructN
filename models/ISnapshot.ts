import { IActiveReality, IReality } from "./IReality";

export interface ISnapshot {
  _id: string;
  date: string;
  progress: number;
  project: string;
  createdAt: string;
  updatedAt: string;
  __v: number;
  structure: string;
  status: string;
  visualization: [];
  reality?: [IReality];
}

// export interface ISnapShotDeatils {
//   _id: string;
//   progress: number;
//   status: string;
//   visualizations?: null[] | null;
//   structure: StructureOrProject;
//   date: string;
//   project: StructureOrProject;
//   createdAt: string;
//   updatedAt: string;
// }
export interface StructureOrProject {
  _id: string;
  name: string;
}
