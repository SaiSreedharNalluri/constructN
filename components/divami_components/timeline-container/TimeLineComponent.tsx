import { useEffect, useState } from "react";
import Moment from "moment";
import Pagination from "@mui/material/Pagination";
import PaginationItem from "@mui/material/PaginationItem";
import { KeyboardArrowDown } from "@mui/icons-material";
import KeyboardDoubleArrowLeftIcon from "@mui/icons-material/KeyboardDoubleArrowLeft";
import KeyboardDoubleArrowRightIcon from "@mui/icons-material/KeyboardDoubleArrowRight";
import { ISnapshot } from "../../../models/ISnapshot";
import CustomCalender from "../custom-datepicker/CustomCalender";
import {
  PaginationStyle,
  SelectedTimeLine,
  TimeLinePagination,
  TimeLineStyleContainer,
} from "./TimeLineComponentStyles";
import dayjs from "dayjs";

interface IProps {
  currentSnapshot: ISnapshot;
  snapshotList: ISnapshot[];
  snapshotHandler: (snapshotData: ISnapshot) => void;
}

const TimeLineComponent: React.FC<IProps> = ({
  currentSnapshot,
  snapshotList,
  snapshotHandler,
}) => {
  const [bottomNav, setBottomNav] = useState(false);
  const [page, setPage] = useState(2);
  const [oldDate, setOldDate] = useState("");
  const [newDate, setNewDate] = useState("");
  const handleChange = (event: React.ChangeEvent<unknown>, value: number) => {
    setPage(value);
  };

  const toggleTimeline = () => {
    setBottomNav(!bottomNav);
  };

  const setCurrentSnapshot = (snapshot: ISnapshot) => {
    snapshotHandler(snapshot);
  };

  // const getSnapshotDate = () => {
  //   if (currentSnapshot) {
  //     return Moment(currentSnapshot.date).format("Do MMM YYYY");
  //   } else {
  //     return "No Reality";
  //   }
  // };

  useEffect(() => {
    setCurrentSnapshot(snapshotList[page - 1]);
  }, [page]);

  useEffect(() => {
    if (snapshotList.length > 0) {
      setOldDate(snapshotList[0].date);
      setNewDate(snapshotList[snapshotList.length - 1].date);
      setPage(snapshotList.length);
    }
  }, [snapshotList]);

  const disableWeekends = (date: any) => {
    const timelineDates: any[] = [];
    if (snapshotList.length) {
      snapshotList.forEach((element, i) => {
        timelineDates.push(dayjs(element.date).format("YYYY-MM-DD"));
      });
    }
    return timelineDates.indexOf(dayjs(date).format("YYYY-MM-DD")) < 0;
  };

  const calenderStyles = {
    sx: {
      "& .css-bkrceb-MuiButtonBase-root-MuiPickersDay-root:not(.Mui-disabled,.Mui-selected)":
        {
          backgroundColor: "rgba(0, 0, 0, 0.20)",
        },
    },
  };

  // console.log(snapshotList, "snaphsot listt");
  return (
    <>
      {snapshotList && snapshotList.length > 0 && (
        <TimeLineStyleContainer>
          <SelectedTimeLine
            style={{ bottom: bottomNav ? "" : 0 }}
            onClick={toggleTimeline}
            data-testid={"selected-timeline"}
          >
            {Moment(currentSnapshot?.date).format("DD MMM YYYY")}
          </SelectedTimeLine>

          {bottomNav ? (
            <div
              data-testid="bottomNav"
              //  className="absolute flex flex-col items-center z-10 top-0 inset-x-0"
            >
              <div
              // className="bg-gray-300 border border-gray-700 rounded duration-300 cursor-pointer"
              >
                {/* <p onClick={toggleTimeline}>{getSnapshotDate()}</p> */}
              </div>
              {snapshotList && currentSnapshot && (
                <TimeLinePagination>
                  <div className=" flex text-sm">
                    <p>{oldDate && Moment(oldDate).format("DD MMM YY")}</p>
                  </div>
                  <PaginationStyle
                    count={snapshotList?.length}
                    page={page}
                    onChange={handleChange}
                    data-testid={"page"}
                    renderItem={(item: any) => (
                      <PaginationItem
                        slots={{
                          previous: KeyboardDoubleArrowLeftIcon,
                          next: KeyboardDoubleArrowRightIcon,
                        }}
                        {...item}
                      />
                    )}
                  />
                  <div className="flex text-sm items-center ml-1 ">
                    <p>{newDate && Moment(newDate).format("DD MMM YY")} </p>
                  </div>
                  {!isNaN(page) ? (
                    <CustomCalender
                      onChange={(e: any) => handleChange(e, 1)}
                      data-testid="calender"
                      shouldDisableDate={disableWeekends}
                      hideTextField
                      data={{
                        disableAll: true,
                        defaultValue: currentSnapshot?.date,
                        // defaultValue: page === 2 ? newDate : oldDate,
                        disableDays: disableWeekends,
                        styles: calenderStyles,
                      }}
                    />
                  ) : null}
                </TimeLinePagination>
              )}
            </div>
          ) : null}
        </TimeLineStyleContainer>
      )}
    </>
  );
};

export default TimeLineComponent;
