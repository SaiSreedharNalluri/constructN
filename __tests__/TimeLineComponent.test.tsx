
import TimeLineComponent from "../components/divami_components/timeline-container/TimeLineComponent";
import { fireEvent, render, screen } from "@testing-library/react";
import { ISnapshot } from "../models/ISnapshot";
const currentSnapshot: ISnapshot = {
    createdAt:"2023-01-24T15:23:40.031Z",
date:"2023-1-06",
progress:-1,
project:"PRJ201897",
reality:[
    {capture:"CAP820581",
    completedDateTime :"2023-01-30T04:55:02.797Z",
    // createdAt :"2023-01-30T04:55:05.466Z",
    job :"JOB820391",
    mode :"Drone Image",
    snapshot   :    "SNP820031",
    status :"active",
    type:"exterior",
    // updatedAt:"2023-01-30T04:55:05.466Z",
    _id :"RLT832872"}
],
status:"Inactive",
structure:"STR536138",
updatedAt:"2023-01-30T04:55:05.466Z",
// visualizations:[]
}
beforeEach(() => {
    render(
        <TimeLineComponent currentSnapshot={currentSnapshot} snapshotList={[currentSnapshot]} snapshotHandler={() => {}} />
      );  
});

it("renders component unchanged", () => {
  const { container } = render(
    <TimeLineComponent currentSnapshot={currentSnapshot} snapshotList={[currentSnapshot]} snapshotHandler={() => {}} />
  );
  expect(container).toMatchSnapshot();
});

it("check the bottom toggle", () => {
    const timeline = screen.getByTestId(/selected-timeline/i);
      fireEvent.click(timeline);
  });
it("check the timeline changes", () => {
      const timeline = screen.getByTestId(/selected-timeline/i);
      fireEvent.click(timeline);
    const page = screen.getByTestId(/page/i);
      fireEvent.click(page);
  });