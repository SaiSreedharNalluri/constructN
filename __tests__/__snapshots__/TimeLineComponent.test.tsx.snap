// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component unchanged 1`] = `
<div>
  <div
    class="css-y7nvji"
  >
    <div
      class="css-99de4n"
      data-testid="selected-timeline"
      style="bottom: 0px;"
    >
      06 Jan 2023
    </div>
  </div>
</div>
`;
